introduction.Rmd

```diff
@@ -9,27 +9,185 @@ library("dplyr")
+
+```{r, error = TRUE}
+exprs <- vars(height / 100, mass + 50)
+
+rlang::eval_tidy(exprs[[1]])
+
+rlang::eval_tidy(exprs[[1]], data = starwars)
+```
+
+To sum up, the distinctive look and feel of data masking UIs requires suspending the normal evaluation of R code. Once captured as quoted code, it can be resumed in a different context. Unfortunately, the delaying of code has a dark side. While it is straightforward to substitute values, it is harder to substitute column names in delayed code. Making indirect references to columns requires learning a bit of new theory and some new tools.
+
+
+## Unquoting code
+
+Data masking functions prevent the normal evaluation of their arguments by quoting them. Once in possession of the blueprints of their arguments, they create a data mask and resume evaluation in this new context. The flip side of delaying code in this way is that making indirect references to columns is only possible by modifying the quoted code before it gets evaluated. We need a surgery operator for blueprints of R code. This is what the `!!` operator[^4] is all about.
+
+[^4]: Pronounced bang bang.
+
+In the world of normal functions, making indirect references is easy. Expressions that yield the same values can be freely interchanged, a property that is sometimes called [referential transparency](https://en.wikipedia.org/wiki/Referential_transparency):
+
+```{r}
+my_function <- function(x) x * 100
+
+my_function(6)
+
+my_function(2 * 3)
+
+a <- 2
+b <- 3
+my_function(a * b)
+```
+
+Because data masking functions evaluate their quoted arguments in a different context, they do not have this property:
+
+```{r, error = TRUE}
+starwars %>% summarise(avg = mean(height, na.rm = TRUE))
+
+value <- mean(height, na.rm = TRUE)
+starwars %>% summarise(avg = value)
+```
+
+Storing a column name in a variable or passing one as function argument requires the tidy eval operator `!!`. This special operator, only available in quoting functions, acts like a surgical operator for modifying blueprints. To understand what it does, it is best to see it in action. The `qq_show()` helper from rlang processes `!!` and prints the resulting blueprint for a computation:
+
+```{r}
+x <- 1
+
+rlang::qq_show(
+  starwars %>% summarise(out = x)
+)
+
+rlang::qq_show(
+  starwars %>% summarise(out = !!x)
+)
+```
+
+As you can observe, `!!` modifies the quoted code by inlining the value of its operand right into the blueprint. What would it take to create an indirect reference to a column name? Passing its name as a string will not produce what you expect:
+
+```{r}
+col <- "height"
+
+rlang::qq_show(
+  starwars %>% summarise(out = sum(!!col, na.rm = TRUE))
+)
+```
+
+This code amounts to taking the mean of a string, something that R will not be happy about:
+
+```{r, error = TRUE}
+starwars %>% summarise(out = sum(!!col, na.rm = TRUE))
+```
+
+To refer to column names inside a blueprint, we need to inline blueprint material. We need **symbols**:
+
+```{r}
+sym(col)
```

Comment:
It feels like usually `sym()`  is usually an anti-pattern (unless you're getting data from shiny). Can we have a different example here?

Reply:
I agree that it is generally an anti-pattern. However, after many discussions with readers of the programming vignette, I think a major cause of confusion was that we tried too hard to hide the concept of symbols. I don't think you can really understand NSE and tidy eval without understanding symbols and how to create them from strings. Also, it's not so problematic when the symbol represents a column name.

I think eventually we need better tools to transform column names to symbols, which should probably live in tidyselect. These would check the symbol is part of the data and issue a standardised error if not.